# Synthetic Data Generation Using RAGAS - RAG Evaluation with LangSmith

In the following notebook we'll explore a use-case for RAGAS' synthetic testset generation workflow!



- 🤝 BREAKOUT ROOM #1
  1. Use RAGAS to Generate Synthetic Data

- 🤝 BREAKOUT ROOM #2
  1. Load them into a LangSmith Dataset
  2. Evaluate our RAG chain against the synthetic test data
  3. Make changes to our pipeline
  4. Evaluate the modified pipeline

SDG is a critical piece of the puzzle, especially for early iteration! Without it, it would not be nearly as easy to get high quality early signal for our application's performance.

Let's dive in!

# 🤝 BREAKOUT ROOM #1

## Task 1: Dependencies and API Keys

We'll need to install a number of API keys and dependencies, since we'll be leveraging a number of great technologies for this pipeline!

1. OpenAI's endpoints to handle the Synthetic Data Generation
2. OpenAI's Endpoints for our RAG pipeline and LangSmith evaluation
3. QDrant as our vectorstore
4. LangSmith for our evaluation coordinator!

Let's install and provide all the required information below!

## Dependencies and API Keys:

### NLTK Import

To prevent errors that may occur based on OS - we'll import NLTK and download the needed packages to ensure correct handling of data.

In [1]:
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to /Users/Micha/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/Micha/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

In [18]:
import os
import getpass

os.environ["LANGCHAIN_TRACING_V2"] = "true"
os.environ["LANGCHAIN_API_KEY"] = getpass.getpass("LangChain API Key:")

We'll also want to set a project name to make things easier for ourselves.

In [3]:
from uuid import uuid4

os.environ["LANGCHAIN_PROJECT"] = f"AIM - SDG - {uuid4().hex[0:8]}"

OpenAI's API Key!

In [4]:
os.environ["OPENAI_API_KEY"] = getpass.getpass("OpenAI API Key:")

## Generating Synthetic Test Data

We wil be using Ragas to build out a set of synthetic test questions, references, and reference contexts. This is useful because it will allow us to find out how our system is performing.

> NOTE: Ragas is best suited for finding *directional* changes in your LLM-based systems. The absolute scores aren't comparable in a vacuum.

### Data Preparation

We'll prepare our data - which should hopefull be familiar at this point since it's our Use-Case Data!

Next, let's load our data into a familiar LangChain format using the `DirectoryLoader`.

In [5]:
from langchain_community.document_loaders import DirectoryLoader
from langchain_community.document_loaders import PyMuPDFLoader


path = "data/"
loader = DirectoryLoader(path, glob="*.pdf", loader_cls=PyMuPDFLoader)
docs = loader.load()

### Knowledge Graph Based Synthetic Generation

Ragas uses a knowledge graph based approach to create data. This is extremely useful as it allows us to create complex queries rather simply. The additional testset complexity allows us to evaluate larger problems more effectively, as systems tend to be very strong on simple evaluation tasks.

Let's start by defining our `generator_llm` (which will generate our questions, summaries, and more), and our `generator_embeddings` which will be useful in building our graph.

### Unrolled SDG

In [6]:
from ragas.llms import LangchainLLMWrapper
from ragas.embeddings import LangchainEmbeddingsWrapper
from langchain_openai import ChatOpenAI
from langchain_openai import OpenAIEmbeddings
generator_llm = LangchainLLMWrapper(ChatOpenAI(model="gpt-4.1-nano"))
generator_embeddings = LangchainEmbeddingsWrapper(OpenAIEmbeddings())

Next, we're going to instantiate our Knowledge Graph.

This graph will contain N number of nodes that have M number of relationships. These nodes and relationships (AKA "edges") will define our knowledge graph and be used later to construct relevant questions and responses.

In [7]:
from ragas.testset.graph import KnowledgeGraph

kg = KnowledgeGraph()
kg

KnowledgeGraph(nodes: 0, relationships: 0)

The first step we're going to take is to simply insert each of our full documents into the graph. This will provide a base that we can apply transformations to.

In [8]:
from ragas.testset.graph import Node, NodeType

### NOTICE: We're using a subset of the data for this example - this is to keep costs/time down.
for doc in docs:
    kg.nodes.append(
        Node(
            type=NodeType.DOCUMENT,
            properties={"page_content": doc.page_content, "document_metadata": doc.metadata}
        )
    )
kg

KnowledgeGraph(nodes: 64, relationships: 0)

Now, we'll apply the *default* transformations to our knowledge graph. This will take the nodes currently on the graph and transform them based on a set of [default transformations](https://docs.ragas.io/en/latest/references/transforms/#ragas.testset.transforms.default_transforms).

These default transformations are dependent on the corpus length, in our case:

- Producing Summaries -> produces summaries of the documents
- Extracting Headlines -> finding the overall headline for the document
- Theme Extractor -> extracts broad themes about the documents

It then uses cosine-similarity and heuristics between the embeddings of the above transformations to construct relationships between the nodes.

In [9]:
from ragas.testset.transforms import default_transforms, apply_transforms

transformer_llm = generator_llm
embedding_model = generator_embeddings

default_transforms = default_transforms(documents=docs, llm=transformer_llm, embedding_model=embedding_model)
apply_transforms(kg, default_transforms)
kg

Applying HeadlinesExtractor:   0%|          | 0/21 [00:00<?, ?it/s]

Applying HeadlineSplitter:   0%|          | 0/64 [00:00<?, ?it/s]

unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to ap

Applying SummaryExtractor:   0%|          | 0/38 [00:00<?, ?it/s]

Property 'summary' already exists in node '907b38'. Skipping!
Property 'summary' already exists in node '373c6c'. Skipping!
Property 'summary' already exists in node '4070b3'. Skipping!
Property 'summary' already exists in node '6e9259'. Skipping!
Property 'summary' already exists in node '8a50a5'. Skipping!
Property 'summary' already exists in node '1b2056'. Skipping!
Property 'summary' already exists in node 'c689b1'. Skipping!
Property 'summary' already exists in node '40ee84'. Skipping!
Property 'summary' already exists in node '6d3e55'. Skipping!
Property 'summary' already exists in node '1b3053'. Skipping!
Property 'summary' already exists in node '83e288'. Skipping!
Property 'summary' already exists in node 'c17daa'. Skipping!
Property 'summary' already exists in node 'c2498b'. Skipping!
Property 'summary' already exists in node '01863a'. Skipping!
Property 'summary' already exists in node '6554ad'. Skipping!
Property 'summary' already exists in node '1af5a3'. Skipping!
Property

Applying CustomNodeFilter:   0%|          | 0/8 [00:00<?, ?it/s]

Applying [EmbeddingExtractor, ThemesExtractor, NERExtractor]:   0%|          | 0/48 [00:00<?, ?it/s]

Property 'summary_embedding' already exists in node '6e9259'. Skipping!
Property 'summary_embedding' already exists in node '8a50a5'. Skipping!
Property 'summary_embedding' already exists in node '373c6c'. Skipping!
Property 'summary_embedding' already exists in node '1b2056'. Skipping!
Property 'summary_embedding' already exists in node 'c17daa'. Skipping!
Property 'summary_embedding' already exists in node '907b38'. Skipping!
Property 'summary_embedding' already exists in node '83e288'. Skipping!
Property 'summary_embedding' already exists in node '4070b3'. Skipping!
Property 'summary_embedding' already exists in node '6d3e55'. Skipping!
Property 'summary_embedding' already exists in node '40ee84'. Skipping!
Property 'summary_embedding' already exists in node 'c689b1'. Skipping!
Property 'summary_embedding' already exists in node '1b3053'. Skipping!
Property 'summary_embedding' already exists in node 'c2498b'. Skipping!
Property 'summary_embedding' already exists in node '01863a'. Sk

Applying [CosineSimilarityBuilder, OverlapScoreBuilder]:   0%|          | 0/2 [00:00<?, ?it/s]

KnowledgeGraph(nodes: 86, relationships: 711)

We can save and load our knowledge graphs as follows.

In [10]:
kg.save("usecase_data_kg.json")
usecase_data_kg = KnowledgeGraph.load("usecase_data_kg.json")
usecase_data_kg

KnowledgeGraph(nodes: 86, relationships: 711)

Using our knowledge graph, we can construct a "test set generator" - which will allow us to create queries.

In [11]:
from ragas.testset import TestsetGenerator

generator = TestsetGenerator(llm=generator_llm, embedding_model=embedding_model, knowledge_graph=usecase_data_kg)

However, we'd like to be able to define the kinds of queries we're generating - which is made simple by Ragas having pre-created a number of different "QuerySynthesizer"s.

Each of these Synthetsizers is going to tackle a separate kind of query which will be generated from a scenario and a persona.

In essence, Ragas will use an LLM to generate a persona of someone who would interact with the data - and then use a scenario to construct a question from that data and persona.

In [12]:
from ragas.testset.synthesizers import default_query_distribution, SingleHopSpecificQuerySynthesizer, MultiHopAbstractQuerySynthesizer, MultiHopSpecificQuerySynthesizer

query_distribution = [
        (SingleHopSpecificQuerySynthesizer(llm=generator_llm), 0.5),
        (MultiHopAbstractQuerySynthesizer(llm=generator_llm), 0.25),
        (MultiHopSpecificQuerySynthesizer(llm=generator_llm), 0.25),
]

#### ❓ Question #1:

What are the three types of query synthesizers doing? Describe each one in simple terms.

#### ✅ Answer #1:

**SingleHopSpecificQuerySynthesizer**
Generates simple questions that can be answered with a single piece of information from one source (simple facts).
*Example:* What is the capital of Paris 😉

**MultiHopAbstractQuerySynthesizer**
Generates abstract questions that need complex reasoning (connecting concepts)
*Example* How do neural networks relate to artificial intelligence?

**MultiHopSpecificQuerySynthesizer**
Generates questions that need multiple specific facts to answer (complex facts)
*Example* What are the advantages of deep learning over traditional machine learning?

Finally, we can use our `TestSetGenerator` to generate our testset!

In [13]:
testset = generator.generate(testset_size=10, query_distribution=query_distribution)
testset.to_pandas()

Generating personas:   0%|          | 0/3 [00:00<?, ?it/s]

Generating Scenarios:   0%|          | 0/3 [00:00<?, ?it/s]

Generating Samples:   0%|          | 0/11 [00:00<?, ?it/s]

user_input  \
0     What does Acemoglu, 2024 say about AI's impact?   
1   How does the AI model Claude compare to ChatGP...   
2           Computer related occupations use ChatGPT?   
3   What role does personal reflection play in und...   
4   How does the prvacy-preserving methdology in C...   
5   how much of chatgpt messages are non work and ...   
6   Based on the data showing that the most common...   
7   Considering the rapid growth of ChatGPT usage ...   
8   Wha US is the US that ChatGPT use in, and how ...   
9   How does the rapid growth of ChatGPT to over 7...   
10  How does the growth of ChatGPT usage by July 2...   

                                   reference_contexts  \
0   [Introduction ChatGPT launched in November 202...   
1   [Table 1: ChatGPT daily message counts (millio...   
2   [Variation by Occupation Figure 23 presents va...   
3   [Conclusion This paper studies the rapid growt...   
4   [<1-hop>\n\nConclusion This paper studies the ...   
5   [<1-hop>\n\nMonth Non-Work (M) (%) Work (M) (%...   
6   [<1-hop>\n\nConclusion This paper studies the ...   
7   [<1-hop>\n\nTable 1: ChatGPT daily message cou...   
8   [<1-hop>\n\nTable 1: ChatGPT daily message cou...   
9   [<1-hop>\n\nConclusion This paper studies the ...   
10  [<1-hop>\n\nConclusion This paper studies the ...   

                                            reference  \
0   The text mentions that Acemoglu, 2024, is one ...   
1   According to the provided context, ChatGPT's d...   
2   Variation by Occupation Figure 23 shows that 5...   
3   The provided context does not mention personal...   
4   The paper introduces a novel privacy-preservin...   
5   In June 2024, non-work messages accounted for ...   
6   The context indicates that Writing is the most...   
7   The context indicates that in the US, over 70%...   
8   The context indicates that the US refers to th...   
9   By July 2025, ChatGPT had been used weekly by ...   
10  By July 2025, ChatGPT had experienced rapid gr...   

                        synthesizer_name  
0   single_hop_specifc_query_synthesizer  
1   single_hop_specifc_query_synthesizer  
2   single_hop_specifc_query_synthesizer  
3   single_hop_specifc_query_synthesizer  
4   multi_hop_abstract_query_synthesizer  
5   multi_hop_abstract_query_synthesizer  
6   multi_hop_abstract_query_synthesizer  
7   multi_hop_specific_query_synthesizer  
8   multi_hop_specific_query_synthesizer  
9   multi_hop_specific_query_synthesizer  
10  multi_hop_specific_query_synthesizer

### Abstracted SDG

The above method is the full process - but we can shortcut that using the provided abstractions!

This will generate our knowledge graph under the hood, and will - from there - generate our personas and scenarios to construct our queries.



In [14]:
from ragas.testset import TestsetGenerator

generator = TestsetGenerator(llm=generator_llm, embedding_model=generator_embeddings)
dataset = generator.generate_with_langchain_docs(docs, testset_size=10)

Applying HeadlinesExtractor:   0%|          | 0/21 [00:00<?, ?it/s]

Applying HeadlineSplitter:   0%|          | 0/64 [00:00<?, ?it/s]

unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to ap

Applying SummaryExtractor:   0%|          | 0/38 [00:00<?, ?it/s]

Property 'summary' already exists in node '5f62f9'. Skipping!
Property 'summary' already exists in node '9e619d'. Skipping!
Property 'summary' already exists in node '585552'. Skipping!
Property 'summary' already exists in node '75a48b'. Skipping!
Property 'summary' already exists in node '3fc6fa'. Skipping!
Property 'summary' already exists in node '7a85bf'. Skipping!
Property 'summary' already exists in node 'f1a0dd'. Skipping!
Property 'summary' already exists in node 'e5c7ad'. Skipping!
Property 'summary' already exists in node '38c096'. Skipping!
Property 'summary' already exists in node '2ea688'. Skipping!
Property 'summary' already exists in node '02cb4d'. Skipping!
Property 'summary' already exists in node 'f4b3b0'. Skipping!
Property 'summary' already exists in node 'e8b8c5'. Skipping!
Property 'summary' already exists in node 'c24658'. Skipping!
Property 'summary' already exists in node '823495'. Skipping!
Property 'summary' already exists in node '3d4517'. Skipping!
Property

Applying CustomNodeFilter:   0%|          | 0/8 [00:00<?, ?it/s]

Applying [EmbeddingExtractor, ThemesExtractor, NERExtractor]:   0%|          | 0/48 [00:00<?, ?it/s]

Property 'summary_embedding' already exists in node '9e619d'. Skipping!
Property 'summary_embedding' already exists in node 'f1a0dd'. Skipping!
Property 'summary_embedding' already exists in node 'e5c7ad'. Skipping!
Property 'summary_embedding' already exists in node '38c096'. Skipping!
Property 'summary_embedding' already exists in node '3fc6fa'. Skipping!
Property 'summary_embedding' already exists in node '5f62f9'. Skipping!
Property 'summary_embedding' already exists in node '585552'. Skipping!
Property 'summary_embedding' already exists in node '7a85bf'. Skipping!
Property 'summary_embedding' already exists in node '75a48b'. Skipping!
Property 'summary_embedding' already exists in node '02cb4d'. Skipping!
Property 'summary_embedding' already exists in node '2ea688'. Skipping!
Property 'summary_embedding' already exists in node 'e8b8c5'. Skipping!
Property 'summary_embedding' already exists in node 'f4b3b0'. Skipping!
Property 'summary_embedding' already exists in node 'c24658'. Sk

Applying [CosineSimilarityBuilder, OverlapScoreBuilder]:   0%|          | 0/2 [00:00<?, ?it/s]

Generating personas:   0%|          | 0/3 [00:00<?, ?it/s]

Generating Scenarios:   0%|          | 0/3 [00:00<?, ?it/s]

Generating Samples:   0%|          | 0/12 [00:00<?, ?it/s]

In [15]:
dataset.to_pandas()

user_input  \
0   Kulveit et al., 2025 what they say about AI ch...   
1   How does OpenAI's development influence user i...   
2   Could you explain the significance of SOC2 cod...   
3   What is the significance of Personal Reflectio...   
4   Considering the rapid growth of ChatGPT's usag...   
5   How does the increasing percentage of non-work...   
6   How do usage patterns and growth across popula...   
7   Considering the rapid growth of ChatGPT's usag...   
8   Based on the findings of Handa et al. (2025), ...   
9   what Handa et al 2025 say about ChatGPT messag...   
10  Based on the data indicating that over 700 mil...   
11  what July 2025 data show ChatGPT use like, and...   

                                   reference_contexts  \
0   [Introduction ChatGPT launched in November 202...   
1   [Table 1: ChatGPT daily message counts (millio...   
2   [Variation by Occupation Figure 23 presents va...   
3   [Conclusion This paper studies the rapid growt...   
4   [<1-hop>\n\nConclusion This paper studies the ...   
5   [<1-hop>\n\nMonth Non-Work (M) (%) Work (M) (%...   
6   [<1-hop>\n\nIntroduction ChatGPT launched in N...   
7   [<1-hop>\n\nIntroduction ChatGPT launched in N...   
8   [<1-hop>\n\nTable 1: ChatGPT daily message cou...   
9   [<1-hop>\n\nTable 1: ChatGPT daily message cou...   
10  [<1-hop>\n\nConclusion This paper studies the ...   
11  [<1-hop>\n\nConclusion This paper studies the ...   

                                            reference  \
0   The context mentions Kulveit et al., 2025 in r...   
1   The context discusses that nearly 80% of all C...   
2   Variation by Occupation Figure 23 presents var...   
3   According to the context, Personal Reflection ...   
4   The data indicates that a significant portion ...   
5   The data shows that non-work messages increase...   
6   The context indicates that ChatGPT's adoption ...   
7   The rapid adoption of ChatGPT, with over 18 bi...   
8   Handa et al. (2025) report that nearly 80% of ...   
9   Handa et al. (2025) report that nearly 80% of ...   
10  The context shows that by July 2025, ChatGPT h...   
11  In July 2025, ChatGPT had over 700 million wee...   

                        synthesizer_name  
0   single_hop_specifc_query_synthesizer  
1   single_hop_specifc_query_synthesizer  
2   single_hop_specifc_query_synthesizer  
3   single_hop_specifc_query_synthesizer  
4   multi_hop_abstract_query_synthesizer  
5   multi_hop_abstract_query_synthesizer  
6   multi_hop_abstract_query_synthesizer  
7   multi_hop_abstract_query_synthesizer  
8   multi_hop_specific_query_synthesizer  
9   multi_hop_specific_query_synthesizer  
10  multi_hop_specific_query_synthesizer  
11  multi_hop_specific_query_synthesizer

We'll need to provide our LangSmith API key, and set tracing to "true".

# 🤝 BREAKOUT ROOM #2

## Task 4: LangSmith Dataset

Now we can move on to creating a dataset for LangSmith!

First, we'll need to create a dataset on LangSmith using the `Client`!

We'll name our Dataset to make it easy to work with later.

In [21]:
from langsmith import Client

client = Client()

dataset_name = "Use Case Synthetic Data - AIE8"

langsmith_dataset = client.create_dataset(
    dataset_name=dataset_name,
    description="Synthetic Data for Use Cases"
)

We'll iterate through the RAGAS created dataframe - and add each example to our created dataset!

> NOTE: We need to conform the outputs to the expected format - which in this case is: `question` and `answer`.

In [22]:
for data_row in dataset.to_pandas().iterrows():
  client.create_example(
      inputs={
          "question": data_row[1]["user_input"]
      },
      outputs={
          "answer": data_row[1]["reference"]
      },
      metadata={
          "context": data_row[1]["reference_contexts"]
      },
      dataset_id=langsmith_dataset.id
  )

## Basic RAG Chain

Time for some RAG!


In [23]:
rag_documents = docs

To keep things simple, we'll just use LangChain's recursive character text splitter!


In [24]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size = 500,
    chunk_overlap = 50
)

rag_documents = text_splitter.split_documents(rag_documents)

We'll create our vectorstore using OpenAI's [`text-embedding-3-small`](https://platform.openai.com/docs/guides/embeddings/embedding-models) embedding model.

In [25]:
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings(model="text-embedding-3-small")

As usual, we will power our RAG application with Qdrant!

In [26]:
from langchain_community.vectorstores import Qdrant

vectorstore = Qdrant.from_documents(
    documents=rag_documents,
    embedding=embeddings,
    location=":memory:",
    collection_name="Use Case RAG"
)

In [27]:
retriever = vectorstore.as_retriever(search_kwargs={"k": 10})

To get the "A" in RAG, we'll provide a prompt.

In [28]:
from langchain.prompts import ChatPromptTemplate

RAG_PROMPT = """\
Given a provided context and question, you must answer the question based only on context.

If you cannot answer the question based on the context - you must say "I don't know".

Context: {context}
Question: {question}
"""

rag_prompt = ChatPromptTemplate.from_template(RAG_PROMPT)

As is usual: We'll be using `gpt-4.1-mini` for our RAG!

In [29]:
from langchain_openai import ChatOpenAI

llm = ChatOpenAI(model="gpt-4.1-mini")

Finally, we can set-up our RAG LCEL chain!

In [30]:
from operator import itemgetter
from langchain_core.runnables import RunnablePassthrough, RunnableParallel
from langchain.schema import StrOutputParser

rag_chain = (
    {"context": itemgetter("question") | retriever, "question": itemgetter("question")}
    | rag_prompt | llm | StrOutputParser()
)

In [31]:
rag_chain.invoke({"question" : "What are people doing with AI these days?"})

'Based on the provided context, people are using AI, particularly generative AI like ChatGPT, in a variety of ways including:\n\n- Performing workplace tasks, either augmenting or automating human labor.\n- Producing writing, software code, spreadsheets, and other digital products.\n- Seeking information and advice, similar to traditional web search engines but with more flexibility.\n- Using AI as co-workers producing output or as co-pilots giving advice and improving human problem-solving productivity.\n- Engaging in activities related to relationships, personal reflection, games, and role play, though these are a smaller share of usage.\n- Therapy and companionship are also noted as prevalent use cases for generative AI.\n\nOverall, generative AI is highly flexible and is used both for professional purposes such as labor and decision-making, and for personal purposes such as self-expression and companionship.'

## LangSmith Evaluation Set-up

We'll use OpenAI's GPT-4.1 as our evaluation LLM for our base Evaluators.

In [32]:
eval_llm = ChatOpenAI(model="gpt-4.1")

We'll be using a number of evaluators - from LangSmith provided evaluators, to a few custom evaluators!

In [33]:
from langsmith.evaluation import LangChainStringEvaluator, evaluate

qa_evaluator = LangChainStringEvaluator("qa", config={"llm" : eval_llm})

labeled_helpfulness_evaluator = LangChainStringEvaluator(
    "labeled_criteria",
    config={
        "criteria": {
            "helpfulness": (
                "Is this submission helpful to the user,"
                " taking into account the correct reference answer?"
            )
        },
        "llm" : eval_llm
    },
    prepare_data=lambda run, example: {
        "prediction": run.outputs["output"],
        "reference": example.outputs["answer"],
        "input": example.inputs["question"],
    }
)

dopeness_evaluator = LangChainStringEvaluator(
    "criteria",
    config={
        "criteria": {
            "dopeness": "Is this response dope, lit, cool, or is it just a generic response?",
        },
        "llm" : eval_llm
    }
)

#### 🏗️ Activity #2:

Highlight what each evaluator is evaluating.

- `qa_evaluator`: evaluates the quality of question-answer pairs -> "How well is our RAG system answering questions" It checks
  - **The answer quality**: How good is the generated answer.
  - **Relevance**: Does the answer actually address the question?
  - **Completeness**: Is the answer complete and informative?
  - **Accuracy**: Is the answer factually correct?

- `labeled_helpfulness_evaluator`: evaluates if the answer is helpful to the user.
- `dopeness_evaluator`: evaluates how "dope" or "cool" the answer is

## LangSmith Evaluation

In [34]:
evaluate(
    rag_chain.invoke,
    data=dataset_name,
    evaluators=[
        qa_evaluator,
        labeled_helpfulness_evaluator,
        dopeness_evaluator
    ],
    metadata={"revision_id": "default_chain_init"},
)

View the evaluation results for experiment: 'fresh-detail-16' at:
https://smith.langchain.com/o/54cd3916-cf32-4735-9053-ad02339e1abb/datasets/d9f59057-5dde-4610-b0b5-85fc4c9f7549/compare?selectedSessions=2f0bbd8a-d506-42ee-ab94-38a8c6171787




0it [00:00, ?it/s]

<ExperimentResults fresh-detail-16>

## Dope-ifying Our Application

We'll be making a few changes to our RAG chain to increase its performance on our SDG evaluation test dataset!

- Include a "dope" prompt augmentation
- Use larger chunks
- Improve the retriever model to: `text-embedding-3-large`

Let's see how this changes our evaluation!

In [35]:
DOPENESS_RAG_PROMPT = """\
Given a provided context and question, you must answer the question based only on context.

If you cannot answer the question based on the context - you must say "I don't know".

Make your answer rad, ensure high levels of dopeness. Do not be generic, or give generic responses.

Context: {context}
Question: {question}
"""

dopeness_rag_prompt = ChatPromptTemplate.from_template(DOPENESS_RAG_PROMPT)

In [36]:
rag_documents = docs

In [44]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size = 500,
    chunk_overlap = 50
)

rag_documents = text_splitter.split_documents(rag_documents)

#### ❓ Question #2:

Why would modifying our chunk size modify the performance of our application?

#### ✅ Answer #2:

By changing the chunk size (in our case: making it bigger), the amount of information per chunk changes. Smaller chunks have more focused information, but relationships between information or their 'context' might get lost. Bigger chunks can preserve relationships between information better, but can also include irrelevant information that dilutes the context.

We have 50% single-hop questions and 50% multi-hop questions. Single-hop questions might perform better with smaller chunk sizes, while multi-hop (especially abstract multi-hop) questions might benefit more from larger chunks. So it is hard to tell if the increase in chunk size will increase the performance or not.

In [38]:
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings(model="text-embedding-3-large")

#### ❓Question #3:

Why would modifying our embedding model modify the performance of our application?

#### ✅ Answer #3:

A better embedding model results in more relevant content being retrieved, which helps the LLM to return better answers. The text-embedding-3-large model has higher dimensional vectors that capture semantic meaning more accurately than text-embedding-3-small. This means it can better understand relationships between concepts and find more relevant chunks when performing similarity searches. So in our case, going from text-embedding-3-small to text-embedding-3-large should definitely improve the performance of our application.

In [45]:
vectorstore = Qdrant.from_documents(
    documents=rag_documents,
    embedding=embeddings,
    location=":memory:",
    collection_name="Use Case RAG Docs"
)

In [46]:
retriever = vectorstore.as_retriever()

Setting up our new and improved DOPE RAG CHAIN.

In [47]:
dopeness_rag_chain = (
    {"context": itemgetter("question") | retriever, "question": itemgetter("question")}
    | dopeness_rag_prompt | llm | StrOutputParser()
)

Let's test it on the same output that we saw before.

In [48]:
dopeness_rag_chain.invoke({"question" : "How are people using AI to make money?"})

"Alright, strap in—the AI money game is more lit than you thought. The context drops some serious knowledge bombs on how ChatGPT and generative AI are flipping the script in the money-making arena.\n\nFirst off, peeps aren’t just using AI to clock in and grind on mundane tasks. Nah, that’s old news. The real sauce is AI acting as a **smart advisor and research assistant**—boosting decision-making, firing up productivity, and unleashing worker output like a productivity beast on steroids.\n\nCollis and Brynjolfsson (2025) put a number on this magic: US users would demand a cool $98 each just to *pause* using generative AI for a month. That’s a jaw-dropping **$97 billion a year in consumer surplus**. Translation? AI’s not just helping people earn—it's creating insane value and economic surplus just by being their confidential sidekick at work and beyond.\n\nSo, the money moves come from:\n- **Augmenting human labor** (not just replacing it)\n- Giving **decision support and research boost

Finally, we can evaluate the new chain on the same test set!

In [49]:
evaluate(
    dopeness_rag_chain.invoke,
    data=dataset_name,
    evaluators=[
        qa_evaluator,
        labeled_helpfulness_evaluator,
        dopeness_evaluator
    ],
    metadata={"revision_id": "dopeness_rag_chain"},
)

View the evaluation results for experiment: 'best-pen-78' at:
https://smith.langchain.com/o/54cd3916-cf32-4735-9053-ad02339e1abb/datasets/d9f59057-5dde-4610-b0b5-85fc4c9f7549/compare?selectedSessions=198733f6-c00c-4339-b4e5-34d249a63a8d




0it [00:00, ?it/s]

<ExperimentResults best-pen-78>

#### 🏗️ Activity #3:

Provide a screenshot of the difference between the two chains, and explain why you believe certain metrics changed in certain ways.

### Analysis:

The system performed better in Test 2 with the larger chunk size and the better embedding model. Especially for 'Helpfulness,' it performed dramatically better (0.4167 vs. 0.833). For correctness, there was no difference in numbers (both had 10 out of 12 correct).

What did change, though, is that in Test 1 there were two 'I don't know' answers, while in Test 2 the answers were actually wrong. This means the system went from being conservative (saying 'I don't know') in Test 1 to being overconfident (giving wrong answers). This suggests the improvements made the system more likely to find information, but not always the correct information.

However, we changed two parameters at the same time: chunk size and the embedding model. This makes it harder to identify which of these changes had the biggest impact. It was to be expected that the results would be better because of the better embedding model, but we could not see what effect the increase in chunk size had. It could also be possible that the results would be even better with the smaller chunk size and the better embedding model.

That is why I did a third evaluation run with the smaller chunk size but the better model. It turned out that the correctness number did not change. We also had two wrong answers (instead of two non-answers), which suggests that the overconfidence was caused by the better embedding model. What did change was the 'Helpfulness' number - down from 0.833 to 0.75. So the chunk size did have a small impact, but not as much as the better model.

'Dopeness' went from 0 to 1, but this was to be expected. This factor should not affect the result quality because it changes just the tone of the response.

| Test       | Chunk Size | Embedding | Correctness | Helpfulness | Dopeness |
| ---------- | ---------- | --------- | ----------- | ----------- | -------- |
| **Test 1** | 500        | 3-small   | 83%         | 33%         | 0%       |
| **Test 2** | 1000       | 3-large   | 83%         | 83%         | 100%      |
| **Test 3** | 500        | 3-large   | 83%         | 75%         | 100%      |


Here are the tests:

Test 1:

![image](img/langsmith-eval-1.png)

Test 2:

![image](img/langsmith-eval-2.png)

Test 3:

![image](img/langsmith-eval-3.png)
